src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs

```diff
@@ -1688,53 +1674,50 @@ internal static class Defaults
-            double initLearningRate = Arguments.Defaults.InitLearningRate,
-            float l2Weight = Arguments.Defaults.L2Weight,
-            ISupportClassificationLossFactory loss = null,
-            Action<Arguments> advancedSettings = null)
+            int maxIterations = Options.Defaults.MaxIterations,
+            double initLearningRate = Options.Defaults.InitLearningRate,
+            float l2Weight = Options.Defaults.L2Weight,
+            ISupportClassificationLossFactory loss = null)
             : base(env, featureColumn, TrainerUtils.MakeBoolScalarLabel(labelColumn), weightColumn)
         {
             Host.CheckNonEmpty(featureColumn, nameof(featureColumn));
             Host.CheckNonEmpty(labelColumn, nameof(labelColumn));
 
-            _args = new Arguments();
-            _args.MaxIterations = maxIterations;
-            _args.InitLearningRate = initLearningRate;
-            _args.L2Weight = l2Weight;
-
-            // Apply the advanced args, if the user supplied any.
-            advancedSettings?.Invoke(_args);
+            _options = new Options();
+            _options.MaxIterations = maxIterations;
+            _options.InitLearningRate = initLearningRate;
+            _options.L2Weight = l2Weight;
 
-            _args.FeatureColumn = featureColumn;
-            _args.LabelColumn = labelColumn;
-            _args.WeightColumn = weightColumn;
+            _options.FeatureColumn = featureColumn;
+            _options.LabelColumn = labelColumn;
+            _options.WeightColumn = weightColumn != null ? Optional<string>.Explicit(weightColumn) : Optional<string>.Implicit(DefaultColumnNames.Weight);
 
             if (loss != null)
-                _args.LossFunction = loss;
-            _args.Check(env);
+                _options.LossFunction = loss;
+            _options.Check(env);
 
-            _loss = _args.LossFunction.CreateComponent(env);
+            _loss = _options.LossFunction.CreateComponent(env);
             Info = new TrainerInfo(calibration: !(_loss is LogLoss), supportIncrementalTrain: true);
-            NeedShuffle = _args.Shuffle;
+            NeedShuffle = _options.Shuffle;
         }
 
         /// <summary>
         /// Initializes a new instance of <see cref="StochasticGradientDescentClassificationTrainer"/>
         /// </summary>
-        internal StochasticGradientDescentClassificationTrainer(IHostEnvironment env, Arguments args)
-            : base(env, args.FeatureColumn, TrainerUtils.MakeBoolScalarLabel(args.LabelColumn), args.WeightColumn)
+        /// <param name="env">The environment to use.</param>
+        /// <param name="options">Advanced arguments to the algorithm.</param>
+        internal StochasticGradientDescentClassificationTrainer(IHostEnvironment env, Options options)
+            : base(env, options.FeatureColumn, TrainerUtils.MakeBoolScalarLabel(options.LabelColumn), options.WeightColumn.IsExplicit ? options.WeightColumn.Value : null)
```

Comment:
I am confused on why the if check is needed. You are saying that if the WeightColumn is explicitly set use the value, otherwise pass null. If null is used when its not set, then is DefaultColumnNames.Weight needed? If not, could the line above:
_options.WeightColumn = ... : Optional<string>.Implicit(DefaultColumnNames.Weight) to:
_options.WeightColumn = ... : Optional<string>.Implicit(null)

To me the code above does the branching to figure out what parameter to use - its weird to branch again to figure out if its explicit or not.

Reply:
couldn't follow the question properly .  

basically without this, some tests fail.  

---
In reply to: [248871449](https://github.com/dotnet/machinelearning/pull/2093#discussion_r248871449) [](ancestors = 248871449)

Reply:
I see what is happening. So Options.WeightColumn is an Optional and it does default to DefaultColumNames.Weight...
Maybe change this line:
     _options.WeightColumn = weightColumn != null ? Optional<string>.Explicit(weightColumn) : Optional<string>.Implicit(DefaultColumnNames.Weight);
to
     _options.WeightColumn = weightColumn != null ? Optional<string>.Explicit(weightColumn) ;

Since its setup with a default?


---
In reply to: [248873304](https://github.com/dotnet/machinelearning/pull/2093#discussion_r248873304) [](ancestors = 248873304,248871449)